Meta: Verify features meta on CI

gamemaker1 opened the pull request:
## Related Issues

Closes #4698 if merged.

## What this PR does

It ensures:
- Only .tsx and .css files exist in `/source/features`.
- Each feature is mentioned in the readme.
- Each file is imported in `refined-github.tsx`.

gamemaker1 commented on package.json:
```diff
@@ -1,6 +1,5 @@
 {
 	"private": true,
-	"type": "module",
```
Not sure if this is the right way to make the script run: one solution is to convert both scripts to JS instead. If `type: module` is mentioned, I get the following error:

```
TypeError: Unknown file extension ".ts" for .../refined-github/build/verify-features.ts
```

Let me know if there is a better way to do this. Thanks

gamemaker1 replied:
Seems running `ava` also fails when I remove `type: module`. I'll look for a better way, let me know if you already know one.

fregante replied:
Yeah this can't be dropped. Probably a solution is in here: https://github.com/TypeStrong/ts-node/issues/1007

I suppose the command is

```sh
node --loader ts-node/esm ./build/verify-features.ts
```

gamemaker1 replied:
Will try that, thanks!

yakov116 commented on build/verify-features.ts:
```diff
@@ -1,52 +1,57 @@
 for (let fileName of featuresDirContents) {
-	const fileExt = getFileExtension(fileName)
+	const fileExt = getFileExtension(fileName);
 	if (fileExt === 'css' || fileName === 'index.tsx' || fileName.includes('rgh')) {
```
```suggestion
	if (('css', 'index.tsx').includes(fileName) || fileName.includes('rgh')) {
```

yakov116 commented on build/verify-features.ts:
```diff
@@ -10,11 +10,11 @@ const featuresInReadme = getFeaturesMeta();
 
 for (let fileName of featuresDirContents) {
-	if (['css', 'index.tsx'].includes(fileName) || fileName.includes('rgh')) {
+	if (fileName === 'index.tsx' || fileName.includes('.css') || fileName.includes('rgh')) {
```
Exactly why I deleted the comment :)

gamemaker1 replied:
Ahh I pressed resolve and the comment disappeared - now I know why 😅

yakov116 replied:
```suggestion
	if (fileName === 'index.tsx' || fileName.endsWith('.css') || fileName.startsWith('rgh')) {
```

yakov116 commented on build/verify-features.ts:
```diff
@@ -0,0 +1,47 @@
+		continue;
+	}
+
+	if (!fileName.includes('.tsx')) {
```
```suggestion
	if (!fileName.endsWith('.tsx')) {
```

fregante commented on .github/workflows/features.yml:
```diff
@@ -1,37 +1,23 @@
-name: Feature
+name: Verify Features
```
```suggestion
name: Features
```

fregante commented on .github/workflows/features.yml:
```diff
@@ -1,37 +1,23 @@
-  New:
-    outputs:
-      name: ${{ steps.name.outputs.first_match }}
+  Verify-Features:
```
```suggestion
  Verify:
```

fregante commented on .github/workflows/features.yml:
```diff
@@ -1,37 +1,23 @@
+          node-version: 16
+          cache: npm
+      - run: npm ci
+      - name: Run the `verify-features` script
+        run: npx ts-node build/verify-features.ts
```
```suggestion
      - run: npx ts-node build/verify-features.ts
```

fregante commented on build/verify-features.ts:
```diff
@@ -0,0 +1,47 @@
+import path from 'node:path';
+import {readdirSync as readDirSync, readFileSync} from 'node:fs';
```
No need to rename code APIs

```suggestion
import {readdirSync, readFileSync} from 'node:fs';
```

fregante commented on build/verify-features.ts:
```diff
@@ -0,0 +1,47 @@
+	}
+}
+
+if (errors.length > 0) {
+	for (const error of errors) {
+		console.error('ERR:', error);
+	}
+
+	throw new Error('Feature verification failed!');
+} else {
+	console.info('All features verified!');
+}
```
```suggestion
for (const error of errors) {
	console.error('ERR:', error);
}

process.exitCode = errors.length;
```

fregante reviewed (commented):
Great to see this so quickly! 🍰

fregante commented on build/verify-features.ts:
```diff
@@ -0,0 +1,47 @@
+import {getFeaturesMeta} from './readme-parser';
+
+const featuresDirContents = readDirSync(path.join(__dirname, '../source/features/'));
+const refinedGithubTs = readFileSync(path.join(__dirname, '../source/refined-github.ts')).toString('utf-8');
```
You can export these two lines in their own functions (and reuse them in `readme-parser` too)

https://github.com/sindresorhus/refined-github/blob/cb722b0639a0bf5f546b036ff95d9f4c26930517/build/readme-parser.ts#L56-L57

```suggestion
const importedFeatures = getImportedFeatures();
```

gamemaker1 commented:
Using `node --loader ts-node/esm build/verify-features.ts` with `type: module` in package.json gives the following error:

```
ReferenceError: exports is not defined in ES module scope
```

gamemaker1 commented:
I got it to work with a bit of a hack - I created a different `tsconfig.json` and added `module: es2020` to it. Running `node --loader ts-node/esm ./build/verify-features.ts` with the `TS_NODE_PROJECT` environment variable set to `./build/tsconfig.json` works!

gamemaker1 commented:
Let me know if you find a better way to do it

Thanks!

gamemaker1 commented:
I tried that, but webpack failed to build the extension

gamemaker1 commented:
I should have mentioned that 😅

gamemaker1 commented:
I tried adding a shebang and the script (but I kept the tsconfig file), let me know if it is ok

gamemaker1 commented:
Yay! Managed to get rid of the extra tsconfig.json by setting the `TS_NODE_COMPILER_OPTIONS` environment variable to `'{"module":"es2020"}'`. Added that to the shebang and deleted `build/tsconfig.json`.

fregante commented:
No it looks great! I was about to push the same config when I saw you figured it out before me 😂

fregante commented on build/readme-parser.ts:
```diff
@@ -53,7 +57,7 @@ export function getFeaturesMeta(): FeatureMeta[] {
 
 export function getFeatures(): FeatureID[] {
-	const contents = readFileSync(path.join(__dirname, '../source/refined-github.ts'), 'utf-8');
+	const contents = getImportedFeatures();
```
I just realized that `getFeatures` already parses the whole file so you can just use that directly instead of creating `getImportedFeatures`

fregante commented on build/verify-features.ts:
```diff
@@ -0,0 +1,44 @@
+
+	const featureMeta = featuresInReadme.find(feature => feature.id === fileName);
+	if (!featureMeta) {
+		errors.push(`readme: The feature ${fileName} is not included in the readme.`);
```
```suggestion
		errors.push(`ERR: ${fileName} should be described in the readme`);
```

fregante commented on build/verify-features.ts:
```diff
@@ -0,0 +1,44 @@
+	}
+}
+
+for (const error of errors) {
+	console.error('ERR:', error);
+}
```
```suggestion
console.error(errors.join('\n'));
```

fregante commented on build/verify-features.ts:
```diff
@@ -0,0 +1,44 @@
+	}
+
+	if (!importedFeatures.includes(featureMeta.id)) {
+		errors.push(`import: The feature ${featureMeta.id} has not been imported in \`/sources/refined-github.ts\`.`);
```
```suggestion
		errors.push(`ERR: ${fileName} should be imported by \`/sources/refined-github.ts\``);
```

fregante commented on build/verify-features.ts:
```diff
@@ -0,0 +1,44 @@
+	}
+
+	if (featureMeta.description.length < 20) {
+		errors.push(`desc: The description for ${featureMeta.id} is less than 20 characters. Try explaining it better!`);
```
```suggestion
		errors.push(`ERR: ${fileName} should be described better in the readme (at least 20 characters)`);
```

fregante commented on build/verify-features.ts:
```diff
@@ -0,0 +1,44 @@
+	}
+
+	if (!fileName.endsWith('.tsx')) {
+		errors.push(`fileext: The \`/source/features\` folder should only contain .css and .tsx files. File \`${fileName}\` violates that rule.`);
```
```suggestion
		errors.push(`ERR: The \`/source/features\` folder should only contain .css and .tsx files. File \`${fileName}\` violates that rule.`);
```

fregante commented on build/verify-features.ts:
```diff
@@ -0,0 +1,44 @@
+		errors.push(`desc: The description for ${featureMeta.id} is less than 20 characters. Try explaining it better!`);
+	}
+
+	if (!importedFeatures.includes(featureMeta.id)) {
```
Let's move this check to the second place and then move this check after:

```js
 	if (fileName.startsWith('rgh-')) {
 		continue;
 	}
```

This way `rgh` features will still be covered by the first 2 rules

gamemaker1 replied:
The RGH features are not supposed to be described in the README right?

gamemaker1 replied:
I didn't understand which checks the `rgh-` files should be covered by

fregante replied:
Done in my last commit, I hope it works.

![borat-picture-a287a51073c07cbc5566aff98c64beb780bde49b](https://user-images.githubusercontent.com/1402241/130597948-fa750a98-c0e8-4188-8477-181f6399c474.jpg)

gamemaker1 replied:
It works, thanks

fregante commented:
<img width="416" alt="Screen Shot" src="https://user-images.githubusercontent.com/1402241/130588198-e9a1ba2a-782a-42eb-8689-763f8a78e634.png">

Please click these buttons to merge the suggestions unless they're incomplete

gamemaker1 commented:
Ok sure 😅

fregante commented on build/verify-features.ts:
```diff
@@ -20,15 +20,15 @@ for (let fileName of featuresDirContents) {
-	if (!importedFeatures.includes(featureMeta.id)) {
-		errors.push(`ERR: ${fileName} should be imported by \`/sources/refined-github.ts\`.`);
+	const featureId = fileName.replace('.tsx', '');
+	if (!importedFeatures.includes(featureId)) {
```
I think it needs a `featureId as FeatureID`

gamemaker1 replied:
Yes that worked

fregante commented on build/verify-features.ts:
```diff
@@ -20,15 +20,15 @@ for (let fileName of featuresDirContents) {
-		errors.push(`ERR: ${fileName} should be imported by \`/sources/refined-github.ts\`.`);
+	const featureId = fileName.replace('.tsx', '');
+	if (!importedFeatures.includes(featureId)) {
+		errors.push(`ERR: ${featureId} should be imported by \`/sources/refined-github.ts\`.`);
```
No period at the end, which would not have been there had you clicked the "accept suggestion" instead 😃

gamemaker1 replied:
Ahh yes sorry about that

gamemaker1 commented:
Interesting but unrelated: I tried reacting with a 😂 emoji to [your message](https://github.com/sindresorhus/refined-github/pull/4699/#issuecomment-904437137) but I couldn't - it only showed me 8 emojis... Is there something I'm missing that allows me to do that?

fregante commented:
GitHub only has 8 reaction emojis 8️⃣🎱🕣♾

fregante commented on build/readme-parser.ts:
```diff
@@ -53,7 +53,7 @@ export function getFeaturesMeta(): FeatureMeta[] {
 
 export function getFeatures(): FeatureID[] {
-	const contents = readFileSync(path.join(__dirname, '../source/refined-github.ts'), 'utf-8');
+	const contents = readFileSync('source/refined-github.ts').toString('utf-8');
```
```suggestion
	const contents = readFileSync('source/refined-github.ts', 'utf-8');
```

gamemaker1 commented:
> GitHub only has 8 reaction emojis eight8ballclock830infinity

Hmmm maybe a possible Refined Github feature?

yakov116 commented:
@gamemaker1 can you remove a readme item to test it's working? (Remove commit and then restore )
Otherwise it looks good to me.

gamemaker1 commented:
Sure

gamemaker1 commented:
It works!

```
$ npm run test:features                                                                                         

> test:features
> ./build/verify-features.ts

(node:7366) ExperimentalWarning: --experimental-loader is an experimental feature. This feature could change at any time
(Use `node --trace-warnings ...` to show where the warning was created)
ERR: The feature copy-on-y should be described in the readme
ERR: view-markdown-source should be described better in the readme (at least 20 characters)
```

yakov116 reviewed (approved):
Perfect! Great Job!

fregante commented:
Could you also ensure that the features are not duplicate in the readme? It's possible that we describe one twice and only ever the first one is read.
